File: src/BoardUtils.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Victor Kwak
 */
public class BoardUtils {
    /**
     * Uses Fisher-Yates shuffle to shuffle array
     *
     * @param toShuffle array to shuffle
     * @return Shuffled array
     */
    public static void shuffle(Square[][] toShuffle) {
        Random random = new Random();
        //Shuffle columns
        for (int j = 0; j < toShuffle[0].length; j++) {
            int rand = random.nextInt(toShuffle.length);
            swap(toShuffle, 0, j, rand, j);
        }
    }

    public static void swap(Square[][] array, int i, int j, int k, int l) {
        Square temp = array[i][j];
        array[i][j] = array[k][l];
        array[i][j].setXY(i, j);
        array[k][l] = temp;
        array[k][l].setXY(k, l);
    }

    public static Board lowestNeighbor(Board current) {
        List<Board> children = generateChildren(current);
        Board lowest = children.get(0);
        for (int i = 1; i < children.size(); i++) {
            if (children.get(i).getHeuristicCost() < lowest.getHeuristicCost()) {
                lowest = children.get(i);
            }
        }

        //Random
        List<Board> lowestChildren = new ArrayList<>();
        int lowestH = lowest.getHeuristicCost();
        for (Board child : children) {
            if (child.getHeuristicCost() == lowestH) {
                lowestChildren.add(child);
            }
        }
        Random random = new Random();
        return lowestChildren.get(random.nextInt(lowestChildren.size()));
    }

    public static List<Board> generateChildren(Board current) {
        List<Board> children = new ArrayList<>();
        for (Square[] squares : current.getBoard()) {
            for (Square square : squares) {
                if (square instanceof Queen) {
                    Square[][] toPassIn = copyBoard(current.getBoard());
                    generateChildrenForQueen(toPassIn, (Queen) square, children);
                }
            }
        }
        return children;
    }

    private static void generateChildrenForQueen(Square[][] board, Queen queen, List<Board> children) {
        if (queen.x != 0) {
            swap(board, queen.x, queen.y, 0, queen.y);
            children.add(new Board(board));
        }
        for (int i = 0; i < board.length - 1; i++) {
            swap(board, i, queen.y, i + 1, queen.y);
            if (i + 1 != queen.x) {
                children.add(new Board(board));
            }
        }
    }

    private static void printBoard(Square[][] board) {
        for (Square[] squares : board) {
            for (Square square : squares) {
                System.out.print(square);
            }
            System.out.println();
        }
    }

    public static Square[][] copyBoard(Square[][] board) {
        Square[][] newBoard = new Square[board.length][board[0].length];
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] instanceof Queen) {
                    newBoard[i][j] = new Queen(i, j);
                } else {
                    newBoard[i][j] = new Empty(i, j);
                }
            }
        }
        return newBoard;
    }
}

File: src/BoardUtils.scala
import scala.annotation.switch
import scala.collection.mutable.ArrayBuffer
import scala.util.Random

/**
  * Created by cary on 2/21/16.
  */
object BoardUtils {

  def shuffle(toShuffle: Array[Array[Square]]) = {
    val rand: Random = scala.util.Random

    for (i <- toShuffle(0).indices) {
      val rand1: Int = rand.nextInt(toShuffle.length)
      swap(toShuffle, 0, i, rand1, i)
    }
  }

  def swap(arr: Array[Array[Square]], i: Int, j: Int, k: Int, l: Int) = {
    val temp: Square = arr(i)(j)
    arr(i)(j) = arr(k)(l)
    arr(i)(j).x = i
    arr(i)(j).y = j
    arr(k)(l) = temp
    arr(k)(l).x = k
    arr(k)(l).y = l
  }

  def lowestNeighbor(current: Board): Board = {
    val children = generateChildren(current)
    var lowest = children(0)

    for (i <- 1 until children.length) {
      if (children(i).getHeuristicCost < lowest.getHeuristicCost) {
        lowest = children(i)
      }
    }

    var lowestChildren = new ArrayBuffer[Board]()
    val lowestH = lowest.getHeuristicCost
    for (i <- children.indices) {
      if (children(i).getHeuristicCost == lowestH) {
        lowestChildren += children(i)
      }
    }

    val rand = new Random()
    lowestChildren(rand.nextInt(lowestChildren.length))
  }

  private def generateChildren(current: Board): ArrayBuffer[Board] = {
    val children = new ArrayBuffer[Board]()
    for (squares <- current.getBoard) {
      for (square <- squares) {
        (square: @switch) match {
          case queen: Queen =>
            val toPassIn = copyBoard(current.getBoard)._1
            for (board <- generateChildrenForQueen(toPassIn, queen)) {
              children += board
            }
          case empty: Empty =>
        }
      }
    }
    children
  }

  private def generateChildrenForQueen(board: Array[Array[Square]], queen: Queen): ArrayBuffer[Board] = {
    val children = new ArrayBuffer[Board]()
    if (queen.x!= 0) {
      swap(board, queen.x, queen.y, 0, queen.y)
      children += new Board(board)
    }
    for (i <- 0 until board.length - 1) {
      swap(board, i, queen.y, i + 1, queen.y)
      if (i + 1 != queen.x) {
        children += new Board(board)
      }
    }
    children
  }

  def copyBoard(board: Array[Array[Square]]): (Array[Array[Square]], Array[Queen]) = {
    val newBoard = Array.ofDim[Square](board.length, board(0).length)
    val newQueenBoard = Array.ofDim[Queen](board.length)
    var queenIndex = 0
    for (i <- board.indices) {
      for (j <- board(0).indices) {
        (board(i)(j): @switch) match {
          case queen: Queen =>
            newBoard(i)(j) = new Queen(i, j)
            newQueenBoard(queenIndex) = newBoard(i)(j).asInstanceOf[Queen]
            queenIndex += 1
          case empty: Empty =>
            newBoard(i)(j) = new Empty(i, j)
        }
      }
    }
    (newBoard, newQueenBoard)
  }

  /**
    *
    * @param function code to run
    * @tparam R return type of the function
    * @return Tuple with the result of function and number of nanoseconds the function took to run.
    */
  def measureRuntime[R](function: => R): (R, Long) = {
    val start = System.nanoTime()
    val result = function //call by name
    val end = System.nanoTime()
    (result, end - start)
  }
}